File: lanxin/src/crypto.ts
import { createHash } from "node:crypto";
import { createDecipheriv } from "node:crypto";

/**
 * Verify Lanxin callback signature.
 * signature = sha1(sort(token, timestamp, nonce, dataEncrypt))
 */
export function verifyLanxinSignature(params: {
  signToken: string;
  timestamp: string;
  nonce: string;
  dataEncrypt: string;
  expectedSignature: string;
}): boolean {
  const { signToken, timestamp, nonce, dataEncrypt, expectedSignature } = params;
  const paramsArr = [signToken, timestamp, nonce, dataEncrypt].sort();
  const concat = paramsArr.join("");
  const hash = createHash("sha1").update(concat).digest("hex");
  return hash === expectedSignature.toLowerCase();
}

/**
 * Decrypt Lanxin callback payload.
 * 官方格式: dataEncrypt = Base64(AES[random(16B) + eventsLen(4B) + orgId + appId + events])
 * 参考: https://developer.lanxin.cn/official/article?article_id=646edbad3d4e4adb7039c163
 *
 * - dataEncrypt: Base64 密文
 * - aesKey: Base64 32 字节密钥（补 "=" 后解码）
 * - 算法: AES-256-CBC, IV = key 前 16 字节
 * - 解密后: 前 20 字节为头部，余下为 orgId + appId + events(JSON)
 * - events 为 JSON，从前 20 字节之后第一个 '{' 起截取到末尾
 */
export function decryptLanxinPayload(params: {
  dataEncrypt: string;
  aesKey: string;
}): string {
  const { dataEncrypt, aesKey } = params;

  const ciphertext = Buffer.from(dataEncrypt, "base64");
  if (ciphertext.length < 16) {
    throw new Error("Lanxin: ciphertext too short");
  }

  const keyBuffer = Buffer.from(aesKey + "=", "base64");
  if (keyBuffer.length < 32) {
    throw new Error("Lanxin: invalid aes key length");
  }

  const key = keyBuffer.subarray(0, 32);
  const iv = key.subarray(0, 16);

  const decipher = createDecipheriv("aes-256-cbc", key, iv);
  decipher.setAutoPadding(true);
  const decrypted = Buffer.concat([decipher.update(ciphertext), decipher.final()]);

  if (decrypted.length < 21) {
    throw new Error("Lanxin: decrypted payload too short");
  }

  // 按官方格式: random(16B) + eventsLen(4B) + orgId + appId + events
  // 官方示例仅返回完整明文，不解析结构；解密后可能含 events 后的 appId 等尾缀
  // 提取第一个完整的 JSON 对象（括号匹配），避免尾缀导致 JSON.parse 失败
  const rest = decrypted.subarray(20).toString("utf-8");
  const jsonStart = rest.indexOf("{");
  if (jsonStart === -1) {
    throw new Error("Lanxin: no JSON object found in decrypted payload");
  }
  const extracted = extractFirstJsonObject(rest.slice(jsonStart));
  if (!extracted) {
    throw new Error("Lanxin: invalid JSON structure in decrypted payload");
  }
  return extracted;
}

/** 提取字符串中第一个完整的 JSON 对象（支持嵌套），用于过滤尾缀 */
function extractFirstJsonObject(str: string): string | null {
  let depth = 0;
  let inString = false;
  let escape = false;
  let quote = "";
  for (let i = 0; i < str.length; i++) {
    const c = str[i];
    if (escape) {
      escape = false;
      continue;
    }
    if (inString) {
      if (c === "\\") escape = true;
      else if (c === quote) inString = false;
      continue;
    }
    if (c === '"' || c === "'") {
      inString = true;
      quote = c;
      continue;
    }
    if (c === "{") {
      depth++;
      continue;
    }
    if (c === "}") {
      depth--;
      if (depth === 0) return str.slice(0, i + 1);
    }
  }
  return null;
}

File: relay/internal/crypto/crypto.go
package crypto

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/sha1"
	"encoding/base64"
	"encoding/hex"
	"fmt"
	"sort"
	"strings"
)

func VerifySignature(signToken, timestamp, nonce, dataEncrypt, expectedSignature string) bool {
	params := []string{signToken, timestamp, nonce, dataEncrypt}
	sort.Strings(params)
	concat := strings.Join(params, "")
	h := sha1.New()
	h.Write([]byte(concat))
	hash := hex.EncodeToString(h.Sum(nil))
	return hash == strings.ToLower(expectedSignature)
}

// DecryptPayload decrypts a Lanxin callback dataEncrypt field.
//
// Binary format after AES decrypt: random(16B) + contentLen(4B) + content
// Content contains orgId + appId prefix followed by JSON.
// We skip the 20-byte header, find the first '{', and extract the first
// complete JSON object (brace-matched) to handle any trailing bytes.
func DecryptPayload(dataEncrypt, aesKey string) (string, error) {
	ciphertext, err := base64.StdEncoding.DecodeString(dataEncrypt)
	if err != nil {
		return "", fmt.Errorf("base64 decode ciphertext: %w", err)
	}
	if len(ciphertext) < 16 {
		return "", fmt.Errorf("ciphertext too short")
	}

	keyBytes, err := base64.StdEncoding.DecodeString(aesKey + "=")
	if err != nil {
		return "", fmt.Errorf("base64 decode key: %w", err)
	}
	if len(keyBytes) < 32 {
		return "", fmt.Errorf("invalid aes key length")
	}

	key := keyBytes[:32]
	iv := key[:16]

	block, err := aes.NewCipher(key)
	if err != nil {
		return "", fmt.Errorf("create cipher: %w", err)
	}

	mode := cipher.NewCBCDecrypter(block, iv)
	decrypted := make([]byte, len(ciphertext))
	mode.CryptBlocks(decrypted, ciphertext)

	padLen := int(decrypted[len(decrypted)-1])
	if padLen < 1 || padLen > aes.BlockSize {
		return "", fmt.Errorf("invalid PKCS7 padding: %d", padLen)
	}
	decrypted = decrypted[:len(decrypted)-padLen]

	if len(decrypted) < 21 {
		return "", fmt.Errorf("decrypted payload too short (%d bytes)", len(decrypted))
	}

	rest := string(decrypted[20:])
	jsonStart := strings.Index(rest, "{")
	if jsonStart == -1 {
		return "", fmt.Errorf("no JSON object found in decrypted payload")
	}

	result, err := extractFirstJsonObject(rest[jsonStart:])
	if err != nil {
		return "", err
	}
	return result, nil
}

// extractFirstJsonObject extracts the first complete JSON object from a string
// using brace depth matching. Handles nested objects and string escaping.
func extractFirstJsonObject(s string) (string, error) {
	depth := 0
	inString := false
	escape := false
	var quote byte

	for i := 0; i < len(s); i++ {
		c := s[i]
		if escape {
			escape = false
			continue
		}
		if inString {
			if c == '\\' {
				escape = true
			} else if c == quote {
				inString = false
			}
			continue
		}
		if c == '"' || c == '\'' {
			inString = true
			quote = c
			continue
		}
		if c == '{' {
			depth++
			continue
		}
		if c == '}' {
			depth--
			if depth == 0 {
				return s[:i+1], nil
			}
		}
	}
	return "", fmt.Errorf("incomplete JSON object")
}
